File: management-domain/Customer-support/Customer-support/Backend/Nodes/support-automation-service/sentiment-analysis-service/src/application/services/sentiment.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { v4 as uuidv4 } from 'uuid';
import Sentiment from 'sentiment';
import natural from 'natural';
import { SentimentAnalysisInputPort } from '@domain/ports/input';
import { EventPublisher } from '@domain/ports/output';
import { SentimentResult, EmotionResult, AnalysisResult } from '@domain/models';
import { SentimentType, Emotion } from '@domain/enums';
import { SentimentAnalyzedEvent } from '@domain/events';

const sentimentAnalyzer = new Sentiment();
const tokenizer = new natural.WordTokenizer();

const emotionLexicon: Record<Emotion, string[]> = {
  [Emotion.JOY]: ['happy', 'joy', 'delighted', 'pleased', 'excited', 'thrilled', 'cheerful', 'glad', 'satisfied', 'love', 'enjoy', 'fun', 'smile'],
  [Emotion.SADNESS]: ['sad', 'unhappy', 'miserable', 'depressed', 'down', 'disappointed', 'heartbroken', 'cry', 'tears', 'hurt', 'pain', 'lonely', 'loss'],
  [Emotion.ANGER]: ['angry', 'furious', 'rage', 'mad', 'outraged', 'annoyed', 'irritated', 'frustrated', 'hostile', 'bitter', 'hate'],
  [Emotion.FEAR]: ['afraid', 'fear', 'scared', 'frightened', 'terrified', 'anxious', 'worried', 'nervous', 'dread', 'panic', 'concerned'],
  [Emotion.DISGUST]: ['disgusted', 'revolting', 'repulsive', 'sick', 'nauseating', 'gross', 'appalled', 'horrible', 'terrible', 'nasty'],
  [Emotion.SURPRISE]: ['surprised', 'shocked', 'astonished', 'amazed', 'stunned', 'startled', 'unexpected', 'wow', 'incredible'],
  [Emotion.NEUTRAL]: ['okay', 'ok', 'fine', 'normal', 'average', 'typical', 'standard'],
};

const intensifiers = ['very', 'really', 'extremely', 'absolutely', 'completely', 'totally'];
const negators = ['not', 'no', 'never', 'none', 'nobody', 'nothing', 'neither', 'barely', 'hardly'];

@Injectable()
export class SentimentService implements SentimentAnalysisInputPort {
  private readonly logger = new Logger(SentimentService.name);

  constructor(private readonly eventPublisher: EventPublisher) {}

  async analyze(text: string, includeEmotions = true): Promise<AnalysisResult> {
    const sentimentResult = await this.analyzeSentiment(text);
    let emotions: EmotionResult | undefined;
    if (includeEmotions) emotions = await this.analyzeEmotions(text);

    const result = AnalysisResult.create({
      id: uuidv4(), text, sentiment: sentimentResult.toPlainObject(),
      emotions: emotions?.toPlainObject(), timestamp: new Date(), createdAt: new Date(),
    });

    await this.eventPublisher.publish(new SentimentAnalyzedEvent({
      eventId: uuidv4(), text, score: sentimentResult.score, sentiment: sentimentResult.sentiment,
    }));

    return result;
  }

  async analyzeSentiment(text: string): Promise<SentimentResult> {
    const result = sentimentAnalyzer.analyze(text);
    let sentiment: SentimentType;
    if (result.score > 0) sentiment = SentimentType.POSITIVE;
    else if (result.score < 0) sentiment = SentimentType.NEGATIVE;
    else sentiment = SentimentType.NEUTRAL;

    const maxScore = result.tokens.length * 5;
    const confidence = maxScore > 0 ? Math.min(Math.abs(result.score) / (maxScore * 0.3), 1) : 0;
    const keywords = [...(result.positive || []), ...(result.negative || [])];

    return SentimentResult.create({ text, sentiment, score: result.score, confidence: Math.round(confidence * 100) / 100, keywords });
  }

  async analyzeEmotions(text: string): Promise<EmotionResult> {
    const tokens = tokenizer.tokenize(text.toLowerCase()) || [];
    const emotionScores: Record<Emotion, number> = { [Emotion.JOY]: 0, [Emotion.SADNESS]: 0, [Emotion.ANGER]: 0, [Emotion.FEAR]: 0, [Emotion.DISGUST]: 0, [Emotion.SURPRISE]: 0, [Emotion.NEUTRAL]: 0 };
    const keywords: Record<Emotion, string[]> = { [Emotion.JOY]: [], [Emotion.SADNESS]: [], [Emotion.ANGER]: [], [Emotion.FEAR]: [], [Emotion.DISGUST]: [], [Emotion.SURPRISE]: [], [Emotion.NEUTRAL]: [] };

    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i];
      const emotion = this.getWordEmotion(token);
      if (emotion) {
        let score = 1;
        let hasNegator = false;
        for (let j = Math.max(0, i - 3); j < i; j++) {
          if (negators.includes(tokens[j])) { hasNegator = true; break; }
          if (intensifiers.includes(tokens[j])) score *= 1.5;
        }
        if (hasNegator) { emotionScores[this.getOppositeEmotion(emotion)] += score * 0.5; }
        else { emotionScores[emotion] += score; keywords[emotion].push(token); }
      }
    }

    const totalScore = Object.values(emotionScores).reduce((s, v) => s + v, 0) || 1;
    for (const e of Object.keys(emotionScores)) emotionScores[e as Emotion] = Math.round((emotionScores[e as Emotion] / totalScore) * 100) / 100;

    let primaryEmotion = Emotion.NEUTRAL;
    let maxScore = 0;
    for (const [e, s] of Object.entries(emotionScores)) { if (s > maxScore) { maxScore = s; primaryEmotion = e as Emotion; } }

    return EmotionResult.create({ text, primaryEmotion, emotions: emotionScores, confidence: maxScore > 0.3 ? maxScore : 0, keywords });
  }

  async batchAnalyze(texts: string[], includeEmotions = false): Promise<{ results: AnalysisResult[]; summary: any }> {
    const results: AnalysisResult[] = [];
    for (const text of texts) results.push(await this.analyze(text, includeEmotions));
    let positive = 0, negative = 0, neutral = 0, totalScore = 0;
    for (const r of results) {
      if (r.sentiment.sentiment === SentimentType.POSITIVE) positive++;
      else if (r.sentiment.sentiment === SentimentType.NEGATIVE) negative++;
      else neutral++;
      totalScore += r.sentiment.score;
    }
    return { results, summary: { total: results.length, positive, negative, neutral, averageScore: results.length > 0 ? Math.round((totalScore / results.length) * 100) / 100 : 0 } };
  }

  async getStats(): Promise<Record<string, unknown>> {
    return { cacheSize: 0, averageConfidence: 0.75 };
  }

  async clearCache(): Promise<number> { return 0; }

  private getWordEmotion(word: string): Emotion | null {
    const lower = word.toLowerCase().replace(/[^a-z]/g, '');
    for (const [emotion, words] of Object.entries(emotionLexicon)) {
      if (words.includes(lower)) return emotion as Emotion;
    }
    return null;
  }

  private getOppositeEmotion(emotion: Emotion): Emotion {
    const opposites: Record<Emotion, Emotion> = { [Emotion.JOY]: Emotion.SADNESS, [Emotion.SADNESS]: Emotion.JOY, [Emotion.ANGER]: Emotion.NEUTRAL, [Emotion.FEAR]: Emotion.NEUTRAL, [Emotion.DISGUST]: Emotion.JOY, [Emotion.SURPRISE]: Emotion.NEUTRAL, [Emotion.NEUTRAL]: Emotion.NEUTRAL };
    return opposites[emotion] || Emotion.NEUTRAL;
  }
}

File: foundation-domain/ai-platform/Backend/Nodes/nlp-service/app/services/sentiment_service.py
"""
Sentiment Analysis Service
Analyzes text sentiment using various NLP models
"""

import logging
from typing import List, Dict, Any, Optional
import numpy as np

logger = logging.getLogger(__name__)


class SentimentResult:
    """Result of sentiment analysis"""
    def __init__(
        self,
        text: str,
        label: str,
        score: float,
        confidence: float,
        probabilities: Optional[Dict[str, float]] = None
    ):
        self.text = text
        self.label = label  # positive, negative, neutral
        self.score = score  # -1 to 1 scale
        self.confidence = confidence  # 0 to 1
        self.probabilities = probabilities or {}


class SentimentService:
    """
    Sentiment analysis service using multiple approaches
    1. Rule-based (VADER-like)
    2. ML-based (scikit-learn)
    3. Deep Learning (transformers)
    """

    # Sentiment lexicon (simplified VADER-like)
    POSITIVE_WORDS = {
        "good", "great", "excellent", "amazing", "wonderful", "fantastic",
        "love", "happy", "pleased", "satisfied", "recommend", "best",
        "awesome", "perfect", "outstanding", "brilliant", "superb",
        "delighted", "impressed", "helpful", "quality", "beautiful",
        "enjoy", "excited", "fortunate", "grateful", "pleasure"
    }

    NEGATIVE_WORDS = {
        "bad", "terrible", "horrible", "awful", "poor", "worst",
        "hate", "disappointed", "sad", "angry", "frustrated", "annoyed",
        "disgusting", "pathetic", "useless", "waste", "boring",
        "unhappy", "dissatisfied", "regret", "avoid", "problem",
        "issue", "complaint", "refund", "slow", "rude", "expensive"
    }

    BOOSTERS = {"very", "really", "extremely", "absolutely", "completely"}
    NEGATIONS = {"not", "no", "never", "neither", "nor", "none"}

    def __init__(self, model_type: str = "rule-based"):
        self.model_type = model_type
        logger.info(f"Initialized SentimentService with model: {model_type}")

    async def analyze_batch(
        self,
        texts: List[str],
        language: str = "en",
        model: str = "rule-based"
    ) -> List[Dict[str, Any]]:
        """
        Analyze sentiment for a batch of texts

        Args:
            texts: List of text strings
            language: Language code
            model: Model type to use

        Returns:
            List of sentiment results
        """
        results = []

        for text in texts:
            if model == "rule-based":
                result = self._analyze_rule_based(text)
            elif model == "ml":
                result = await self._analyze_ml(text)
            else:
                result = self._analyze_rule_based(text)  # Default to rule-based

            results.append({
                "text": text[:200] + "..." if len(text) > 200 else text,
                "label": result.label,
                "score": result.score,
                "confidence": result.confidence,
                "probabilities": result.probabilities
            })

        return results

    def _analyze_rule_based(self, text: str) -> SentimentResult:
        """Rule-based sentiment analysis (VADER-like approach)"""
        text_lower = text.lower()
        words = text_lower.split()

        # Count positive and negative words
        positive_count = sum(1 for word in words if word in self.POSITIVE_WORDS)
        negative_count = sum(1 for word in words if word in self.NEGATIVE_WORDS)

        # Apply negation
        negated = False
        for i, word in enumerate(words):
            if word in self.NEGATIONS:
                negated = True
                # Check next word
                if i + 1 < len(words):
                    next_word = words[i + 1]
                    if next_word in self.POSITIVE_WORDS:
                        positive_count -= 1
                        negative_count += 1
                    elif next_word in self.NEGATIVE_WORDS:
                        negative_count -= 1
                        positive_count += 1

        # Apply boosters
        for word in words:
            if word in self.BOOSTERS:
                positive_count = positive_count * 1.5
                negative_count = negative_count * 1.5

        # Calculate score
        total_words = len(words)
        if total_words == 0:
            return SentimentResult(text, "neutral", 0.0, 0.0)

        # Normalize score between -1 and 1
        raw_score = (positive_count - negative_count) / max(total_words * 0.1, 1)
        score = max(-1.0, min(1.0, raw_score))

        # Determine label
        if score > 0.1:
            label = "positive"
        elif score < -0.1:
            label = "negative"
        else:
            label = "neutral"

        # Calculate confidence based on extremity
        confidence = abs(score)

        # Calculate probabilities
        pos_prob = max(0, min(1, (score + 1) / 2))
        neg_prob = max(0, min(1, (1 - score) / 2))
        neu_prob = 1 - (pos_prob + neg_prob) / 2

        return SentimentResult(
            text=text,
            label=label,
            score=round(score, 4),
            confidence=round(confidence, 4),
            probabilities={
                "positive": round(pos_prob, 4),
                "negative": round(neg_prob, 4),
                "neutral": round(neu_prob, 4)
            }
        )

    async def _analyze_ml(self, text: str) -> SentimentResult:
        """ML-based sentiment analysis"""
        # This would use a trained ML model
        # For now, fall back to rule-based
        return self._analyze_rule_based(text)

    def _preprocess_text(self, text: str) -> str:
        """Preprocess text for analysis"""
        # Basic preprocessing
        text = text.strip()
        # Remove URLs
        import re
        text = re.sub(r'http\S+', '', text)
        # Remove mentions
        text = re.sub(r'@\w+', '', text)
        # Remove extra whitespace
        text = ' '.join(text.split())
        return text

    def analyze_with_emotion(self, text: str) -> Dict[str, Any]:
        """
        Analyze sentiment with emotion detection

        Returns sentiment plus detected emotions:
        - joy, sadness, anger, fear, surprise, disgust
        """
        text_lower = text.lower()

        # Emotion keywords
        emotions = {
            "joy": ["happy", "excited", "delighted", "thrilled", "joy", "cheerful"],
            "sadness": ["sad", "unhappy", "depressed", "down", "disappointed"],
            "anger": ["angry", "furious", "mad", "irritated", "annoyed"],
            "fear": ["afraid", "scared", "frightened", "anxious", "worried"],
            "surprise": ["surprised", "shocked", "amazed", "astonished"],
            "disgust": ["disgusted", "revolted", "sickened", "appalled"]
        }

        detected_emotions = {}
        for emotion, keywords in emotions.items():
            count = sum(1 for word in keywords if word in text_lower)
            if count > 0:
                detected_emotions[emotion] = min(1.0, count * 0.3)

        sentiment_result = self._analyze_rule_based(text)

        return {
            "sentiment": {
                "label": sentiment_result.label,
                "score": sentiment_result.score,
                "confidence": sentiment_result.confidence
            },
            "emotions": detected_emotions or {"neutral": 1.0}
        }
